### PR TITLE
add `minute_offset` and update to fsrs 0.6.2

### LINKED ISSUE #22: [BUG] The hard-coded timezone in anki_to_fsrs
https://github.com/open-spaced-repetition/fsrs-rs/blob/16bdf86c5053038dd6fc9bf939f395b889be9fe2/src/convertor_tests.rs#L178-L190

@AlexErrant, this function is used by fsrs-browser. It causes some problems for people in other timezone.

https://github.com/open-spaced-repetition/fsrs-browser/blob/58f27babb2bf6aea9de06820156a8213aa1712ff/sandbox/src/train.ts#L138

https://github.com/open-spaced-repetition/fsrs-browser/blob/58f27babb2bf6aea9de06820156a8213aa1712ff/src/lib.rs#L32-L44

### PR DESCRIPTION
**Pull request by AlexErrant**
Closes https://github.com/open-spaced-repetition/fsrs-browser/issues/22

---

Answering some questions from [here](https://github.com/open-spaced-repetition/fsrs-browser/issues/22#issuecomment-2081098173): 
> The minute_offset is not very readable.

Agreed.

> Could you provide an example or document?

Please let me know if this PR's example/documentation is insufficient.

> If the previous next_day_starts_at is 4 and the timezone is UTC+8:00, what's the minute_offset for current method?

I _kinda_ show this in `train.ts`. But to be explicit:

```
minute_offset = utc_offset - next_day_starts_at
minute_offset = 8hr - 4hr
minute_offset = 240 minutes
```

### CODE REVIEW

File: sandbox/src/train.ts [hunk 1]
```diff
@@ -135,7 +135,25 @@ function computeParameters(
...
+		// We subtract ankiNextDayStartsAtMinutes because
+		// we want a review done at, say, 1am to be done for the *prior* day.
+		// Adding would move it into the future.
+		timeZoneMinutesOffset - ankiNextDayStartsAtMinutes,
+		cids,
+		eases,
+		ids,
+		types,
+		progress,
+	)
```
Comment (AlexErrant) [inline sandbox/src/train.ts:156]: Please let me know if the logic in these comments is confusing or wrong!

File: sandbox/src/train.ts [hunk 2]
```diff
@@ -135,7 +135,25 @@ function computeParameters(
...
```
Comment (AlexErrant) [inline sandbox/src/train.ts:156]: Please let me know if the logic in these comments is confusing or wrong!

**Comment by L-M-Sherlock** [comment]
@sobjornstad is using our package in their product (@remnoteio),  so I'd like to ask for his advice.

File: sandbox/src/train.ts [hunk 3]
```diff
@@ -135,7 +135,25 @@ function computeParameters(
...
+	// This is useful when going from our timezone to UTC; our local time + offset = UTC.
+	// However, we want to go from UTC to our timezone. (Anki revlog id is UTC, and we want to convert it to local time.)
+	// So we flip the sign of `getTimezoneOffset()`.
+	let timeZoneMinutesOffset = new Date().getTimezoneOffset() * -1
+	let ankiNextDayStartsAtMinutes = 4 * 60
+	let parameters = fsrs.computeParametersAnki(
+		// We subtract ankiNextDayStartsAtMinutes because
+		// we want a review done at, say, 1am to be done for the *prior* day.
+		// Adding would move it into the future.
+		timeZoneMinutesOffset - ankiNextDayStartsAtMinutes,
```
Comment (ishiko732) [inline sandbox/src/train.ts:150]: It looks fine. If my timezone is `UTC` with a `utc_offset = 0` and `next_day_starts_at = 4`, then the `minute_offset = 240 minutes`.  find the relevant code and context at this link https://github.com/open-spaced-repetition/fsrs-rs/blame/14a62849af95a0df0f97cc1251390d03a1dda19c/src/convertor_tests.rs#L78-L82

So, if I review at `2024-04-28 05:00`, the converted time would be `2024-04-28 01:00`. And if I review at `2024-04-28 03:00`, the converted time would be `2024-04-27 23:00`. Is that correct?

**Review by L-M-Sherlock — APPROVED** [review]
LGTM

**Review by ishiko732 — APPROVED** [review]
LGTM

File: sandbox/src/train.ts (hunk 3 above)
Comment (L-M-Sherlock) [reply sandbox/src/train.ts:150]: Yeah, it's intended. Ideally, we hope the user doesn't do their reviews around the `NextDayStarts`, when they should be sleeping.

Comment (AlexErrant) [reply sandbox/src/train.ts:150]: > If my timezone is UTC with a `utc_offset = 0` and `next_day_starts_at = 4`, then the `minute_offset = 240 minutes`.

To be exceedingly pedantic, `minute_offset = -240 minutes`; i.e. it's negative. This can be seen in [the original code](https://github.com/open-spaced-repetition/fsrs-rs/blame/64b5f518cff07051228461d4275f9501382a4ae3/src/convertor_tests.rs#L114) where `next_day_starts_at` is subtracted. In this PR's code it's changed to addition, so the subtraction must be done client-side (in Javascript) as explained in the comments.

The rest is correct.

**Comment by L-M-Sherlock** [comment]
Do you have a plan for the next release?

**Comment by AlexErrant** [comment]
Yup gimme a sec!

**Comment by AlexErrant** [comment]
I'm not sure if sobjornstad/Remnote uses the Anki revlog format. This PR is only relevant to those using that style; using raw `fsrs-items` means you get to control the timestamps/card-grouping however you wish.

(Editorial time: Honestly, using raw `fsrs-items` is probably the best way to do things. There are some drawbacks to using Anki's format. For example, if you went to Europe for summer vacation, then went home in autumn, each timezone shift would change your training data's [`date_naive()`](https://github.com/open-spaced-repetition/fsrs-rs/blob/14a62849af95a0df0f97cc1251390d03a1dda19c/src/convertor_tests.rs#L81). This is annoying, but probably doesn't significantly effect training. It just bothers some people that their heatmaps are now missing spots haha.)

**Comment by L-M-Sherlock** [comment]
Yeah. Timezone shift is hard to deal with if we use the revlog format. I don't know how Anki deal with that. @dae, sorry for bothering you. Has Anki solved this issue? If so, how?

**Comment by dae** [comment]
Anki revlogs do not record timezone info, so Anki is not aware of historical timezone changes.

**Comment by L-M-Sherlock** [comment]
Fine. Thanks for that information.

**Comment by sobjornstad** [comment]
Sorry for the delay, I didn't check my Discord messages for a while :)

What is the “Anki revlog format” in the context of FSRS? Some choice in the API used? We use `FSRS.repeat()` to replay RemNote's history format onto an FSRS card object, and we pass in the UTC time that the review was conducted. No time zone is recorded in our review logs either.

**Comment by AlexErrant** [comment]
Ah, "Anki revlog format" in the context of `fsrs-browser` means data that's more or less shaped like [the revlog table in Anki.](https://github.com/ankidroid/Anki-Android/wiki/Database-Structure#review-log) To be specific, the `computeParametersAnki` and `memoryStateAnki` methods in [`lib.rs`](https://github.com/open-spaced-repetition/fsrs-browser/blob/26d8744461e4561698aca8035b2aa47227697e56/src/lib.rs#L33) (which are exposed/exported to Javascript) require `cids`, `eases`, `ids`, and `types`.

I think `FSRS.repeat()` is a `py-fsrs` thing? Since `fsrs-browser` is based on `fsrs-rs` which doesn't have a `repeat`, perhaps this discussion is entirely irrelevant to your interests haha.

**Comment by sobjornstad** [comment]
We use `ts-fsrs`, not `py-fsrs`. The only factors involved in using `repeat` are the date of the review and the rating provided.

**Comment by L-M-Sherlock** [comment]
@sobjornstad, this PR is related to the optimization, so you need to check that part of RemNote rather than scheduler.

**Comment by sobjornstad** [comment]
Ah gotcha, I have actually not been involved in the optimizer at all yet. I'll review that and check with Martin tomorrow
